### PR TITLE
NetCDF thread safety take two

### PR DESCRIPTION
**Pull request by trexfeathers**
### Description

Taking the lead from #5061, I have minimised the footprint of this change, making it hopefully easier for future maintenance and better suited to a patch release (our goal).

~~I still get very occasional segfaults and `invalid ID` NetCDF errors. These were guaranteed before, but I would now estimate they fail a full test run ~ 1 in 20 times. This of course makes them very difficult to investigate - I don't know if there are any commits in this PR's sequence where there was genuinely no problem, or whether it was just luck. The only working theory I have is that they show up when certain combinations of tests are being run on separate processors, but I don't know how I would test such a theory.~~

I'm putting this up now because I need help with what's left:

- [x] Investigate remaining intermittent errors (see above) and decide if they are a problem.
- [x] Write test(s) specifically aimed at thread safety.
- [x] Arrange for some real world user testing.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by trexfeathers** [comment]
>Investigate remaining intermittent errors (see above) and decide if they are a problem.

Local testing shows that this was 100% fine at 2fb274f, so it's something I've introduced since.

**Comment by trexfeathers** [comment]
>Investigate remaining intermittent errors (see above) and decide if they are a problem.

Solved by 01be381.

**Comment by trexfeathers** [comment]
@valeriupredoi a lot of the errors you reported in #5016 are consistent with using an older version of Iris. Are you sure you checked out this branch?

**Comment by valeriupredoi** [comment]
massive facepalm! I am using the `master` branch from your fork - which branch to co?

**Comment by valeriupredoi** [comment]
good man! On it now, smelled like really old stuff to me too :laughing:

**Comment by valeriupredoi** [comment]
> > massive facepalm! I am using the `master` branch from your fork - which branch to co?
> 
> [`trexfeathers:netcdf_segs`](https://github.com/trexfeathers/iris/tree/netcdf_segs)

@trexfeathers - very promising - first full run of ESMValCore tests and 0 SegFaults, all tests passed fine :partying_face: Will run a bunch of them, just in case

**Comment by valeriupredoi** [comment]
I am not getting any SegFaults! Repeated the test suite run four times (single proc, I still want to run with `-n 4` or so) but I am getting this hiccup only for the 13 sample data tests we have (they generate pickled netCDF4 data) - seems like a trivial fix at your end to me:
```
tests/sample_data/multimodel_statistics/test_multimodel.py:335: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/sample_data/multimodel_statistics/test_multimodel.py:163: in multimodel_test
    result = multi_model_statistics(products=cubes,
esmvalcore/preprocessor/_multimodel.py:547: in multi_model_statistics
    return _multicube_statistics(
esmvalcore/preprocessor/_multimodel.py:441: in _multicube_statistics
    result_cube = _compute_eager(aligned_cubes,
esmvalcore/preprocessor/_multimodel.py:354: in _compute_eager
    _ = [cube.data for cube in cubes]  # make sure the cubes' data are realized
esmvalcore/preprocessor/_multimodel.py:354: in <listcomp>
    _ = [cube.data for cube in cubes]  # make sure the cubes' data are realized
../iris/lib/iris/cube.py:2436: in data
    return self._data_manager.data
../iris/lib/iris/_data_manager.py:206: in data
    result = as_concrete_data(self._lazy_array)
../iris/lib/iris/_lazy_data.py:280: in as_concrete_data
    (data,) = _co_realise_lazy_arrays([data])
../iris/lib/iris/_lazy_data.py:243: in _co_realise_lazy_arrays
    computed_arrays = da.compute(*arrays)
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/base.py:600: in compute
    results = schedule(dsk, keys, **kwargs)
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/threaded.py:89: in get
    results = get_async(
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/local.py:511: in get_async
    raise_exception(exc, tb)
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/local.py:319: in reraise
    raise exc
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/local.py:224: in execute_task
    result = _execute_task(task, data)
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:119: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:119: in <genexpr>
    return func(*(_execute_task(a, cache) for a in args))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:119: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:119: in <genexpr>
    return func(*(_execute_task(a, cache) for a in args))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:119: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:119: in <genexpr>
    return func(*(_execute_task(a, cache) for a in args))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:119: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/optimization.py:990: in __call__
    return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:149: in get
    result = _execute_task(task, cache)
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/core.py:119: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/utils.py:71: in apply
    return func(*args, **kwargs)
../miniconda3/envs/iris-tests/lib/python3.10/site-packages/dask/array/core.py:120: in getter
    c = a[b]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <NetCDFDataProxy shape=(1980, 2, 3, 2) dtype=dtype('float32') path='/home/valeriu/miniconda3/envs/iris-tests/lib/pytho...1/historical/r1i1p1f1/Amon/ta/gn/v20200623/ta_Amon_TaiESM1_historical_r1i1p1f1_gn_185001-201412.nc' variable_name='ta'>
keys = (slice(0, 1980, None), slice(0, 2, None), slice(0, 3, None), slice(0, 2, None))

    def __getitem__(self, keys):
        # Dask will call this many times during reading, and also provides
        #  dedicated API for lock acquisition, so we use that instead of using
        #  a DatasetContainer ( see iris._lazy_data.as_lazy_data() ).
>       assert self.required_lock.locked()
E       AssertionError

../iris/lib/iris/fileformats/netcdf/_thread_safe.py:319: AssertionError
```
(the reason I didn't get this error the first time is that the files were cached the first time I ran pytest, afterwards, with the second run, they got recreated)

**Comment by valeriupredoi** [comment]
OK more than 10 test sessions for ESMValCore (in single proc and `-n 2`) and still no SegFault - looks like you cured it @trexfeathers :medal_military: All tests pass, bar those sample data ones failing with the assertion error mentioned above. Here is me env, for the record
[iris_tests.yml.txt](https://github.com/SciTools/iris/files/10196210/iris_tests.yml.txt)

:beer:

**Comment by trexfeathers** [comment]
@valeriupredoi those tests did actually reveal a flaw in what I was doing. Please could you try this new version when you are able?

**Comment by valeriupredoi** [comment]
> @valeriupredoi those tests did actually reveal a flaw in what I was doing. Please could you try this new version when you are able?

:beer: for fixing that! Absolutelym will test again later today, and post here, mate :+1:

**Review by pp-mo — CHANGES_REQUESTED** [review]
Ok I'm going to submit these ideas as they are.
I do think the approach is is sound, and we have now some good evidence that it works.

Outanding:
  * decide on additional testing needed  -- I am working on it
  * re-consider the [use of a DatasetContainer in data proxies](https://github.com/SciTools/iris/pull/5095#discussion_r1053394685), and understand why it seems to fail  -- deferred until other changes have stabilised

File: lib/iris/fileformats/netcdf/loader.py [hunk 1]
```diff
@@ -34,6 +33,7 @@
 import iris.coords
 import iris.exceptions
 import iris.fileformats.cf
+from iris.fileformats.netcdf import _thread_safe
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/loader.py:36]: Agreed - a better signal to devs wherever it is referenced without its full path.

File: lib/iris/fileformats/netcdf/_thread_safe.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+import numpy as np
+
+_GLOBAL_NETCDF4_LOCK = Lock()
+
+# Doesn't need thread protection, but this allows all netCDF4 refs to be
+#  replaced with thread_safe refs.
+default_fillvals = netCDF4.default_fillvals
+
+
+class _ThreadSafeAggregator(ABC):
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py:26]: Thanks for doing the leg work on this. I was basing my name on a single Stack Overflow answer.

File: lib/iris/fileformats/netcdf/_thread_safe.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+            with _GLOBAL_NETCDF4_LOCK:
+                instance = self.CONTAINED_CLASS(*args, **kwargs)
+
+        self.__contained = instance
+
+    def __getattr__(self, item):
+        if item[-11:] == "__contained":
+            # Special behaviour when accessing the __contained instance itself.
+            return ABC.__getattribute__(
+                self, f"{_ThreadSafeAggregator.__name__}__contained"
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py:62]: Big plus points for being more Pythonic and more readable. Passes all the Iris tests when I  run locally, although I'll use a two-word name just to fully minimise possible clashes with NetCDF4.

File: lib/iris/fileformats/netcdf/_thread_safe.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+        The original returned netCDF4.Dimensions are simply replaced with their
+        respective DimensionContainers, ensuring that downstream calls are
+        also performed within _GLOBAL_NETCDF4_LOCK.
+        """
+        with _GLOBAL_NETCDF4_LOCK:
+            dimensions_ = self.__contained.get_dims(*args, **kwargs)
+        return tuple(
+            [DimensionContainer._from_existing(d) for d in dimensions_]
+        )
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py:140]: Good catch, thanks

File: lib/iris/tests/test_netcdf.py [hunk 5]
```diff
@@ -19,7 +19,6 @@
 import tempfile
 from unittest import mock
 
-import netCDF4 as nc
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/test_netcdf.py:22]: I get where you're coming from - the diff is quite large and repetitive. But I came up with this scheme specifically to make clear to developers what they're working with, to minimise possible nasty confusions. So, in answer to:

>Do you think this is worth it, to minimise code changes ?

No, in my opinion 😁

File: lib/iris/tests/test_cf.py [hunk 6]
```diff
@@ -52,11 +54,15 @@ def test_cached(self):
...
-    def setUp(self):
+    @pytest.fixture(autouse=True)
+    def set_up(self):
         filename = tests.get_data_path(
             ("NetCDF", "rotated", "xyt", "small_rotPole_precipitation.nc")
         )
         self.cfr = cf.CFReader(filename)
+        with self.cfr:
+            yield
+        pass
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/test_cf.py]: 128ebaa3879288f199fc404ecac066f9bf82ac32

File: lib/iris/fileformats/netcdf/loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/loader.py]: 8d72578fb5f26f6fc718c95db3f8994b7c9082df

File: lib/iris/fileformats/netcdf/_thread_safe.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py]: 68588b50b1f423910ca6f54229f91bce8095388c

File: lib/iris/fileformats/netcdf/_thread_safe.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+    def __getitem__(self, item):
+        with _GLOBAL_NETCDF4_LOCK:
+            return self.__contained.__getitem__(item)
+
+    def __setitem__(self, key, value):
+        with _GLOBAL_NETCDF4_LOCK:
+            return self.__contained.__setitem__(key, value)
+
+
+class DimensionContainer(_ThreadSafeAggregator):
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py]: 68588b50b1f423910ca6f54229f91bce8095388c

File: lib/iris/fileformats/netcdf/_thread_safe.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py]: 5d351db88ab5b0e36374f767552d9ef447b85709

File: lib/iris/fileformats/netcdf/_thread_safe.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+"""
+Module to ensure all calls to the netCDF4 library are thread-safe.
+
+Intention is that no other Iris module should import the netCDF module.
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py]: d53d194cd343cb3cfae89539eb92ba5986ae16bf

File: lib/iris/fileformats/netcdf/_thread_safe.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+
+class _ThreadSafeAggregator(ABC):
+    """
+    Contains a netCDF4 class instance, ensuring wrapping all API calls within _GLOBAL_NETCDF4_LOCK.
+
+    Designed to 'gate keep' all the instance's API calls, but allowing the
+    same API as if working directly with the instance itself.
+
+    Using an aggregator because we cannot successfully subclass or monkeypatch
+    netCDF4 classes, as they are only wrappers for the C-layer.
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py]: d9bce6bc08da73118ec13be3eac555baf22cd0a3

File: lib/iris/fileformats/netcdf/_thread_safe.py [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    """
+
+    CONTAINED_CLASS = NotImplemented
+
+    # Allows easy assertions, without difficulties with isinstance and mocking.
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py]: ed17c8317b95e10d6a1b9322eb13289ce4522908

File: lib/iris/fileformats/netcdf/_thread_safe.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/fileformats/netcdf/_thread_safe.py]: 9b23b30722a69ee9f28e59d92cf3b95c2972c336

As I recall from offline conversation: nothing needed to be done with the members that returned dictionaries, only those that might EITHER return lists OR generators.

**Comment by trexfeathers** [comment]
OK  @pp-mo, those are all my responses.

If the branch update commits are confusing for your review workflow, and you'd prefer me to rebase-and-force-push, just let me know.

**Comment by pp-mo** [comment]
**Update:** I've at last reviewed + OK-d all the changes in response to the initial review, resolved all outstanding conversations.
Discussed testing approach offline with @trexfeathers

We agreed we can't _directly_ address the thread safety problem in testing -- we want to run exercises against the pre-existing netcdf code, and see them fail, but we can't expect to write anything that will _always reliably fail_.
So, write tests to exercise the code in specific ways, and  show that they _sometimes_ fail with the previous code.  This gives confidence, if those failures never happen with the new ( and its re-checked on every PR).

@trexfeathers  to write some tests + demonstrate they do (did) fail sometimes.
Once agreed, we can merge !

**Comment by valeriupredoi** [comment]
that's brilliant work, guys! Very much looking forward to seeing this in, both you and us we'll be happy to see the repopatch-pin on `netCDF4` go, and finally, getting support for Python=3.11 :beer: 

> We agreed we can't directly address the thread safety problem in testing -- we want to run exercises against the pre-existing netcdf code, and see them fail, but we can't expect to write anything that will always reliably fail.

If you guys plug in a few tests a la the ones we have in ESMValTool that were showing the symptoms of the issue, then I reckon all's good - it becomes a case of "so long those tests never fail, we're in business" vs "we need tests that must fail at all times"

**Comment by trexfeathers** [comment]
>If you guys plug in a few tests a la the ones we have in ESMValTool that were showing the symptoms of the issue, then I reckon all's good - it becomes a case of "so long those tests never fail, we're in business" vs "we need tests that must fail at all times"

We do have some tests that reliably fail, but one of our big concerns is that they're not written to test thread safety - so future devs are free to change/remove them without realising it would remove our coverage for thread safety. Hence writing some dedicated tests.

@valeriupredoi are your tests written to test thread safety, or was that an unexpected bonus?

**Comment by valeriupredoi** [comment]
@trexfeathers oh no we don't have such fancy tests designed to test thread safety - just the bog standard netCDF file loading with iris ones, but those are enough to catch the issues that this PR fixes, on a statistical basis though ie the more those tests are run the more SegFaults are seen - you see, we are a bit Soviet here, quantity has its own quality :rofl:

**Review by pp-mo — CHANGES_REQUESTED** [review]
This looks generally like a neat way of doing it.
I think we need to record test failures somewhere, maybe just on this ticket ?

I wonder if these can show errors against netCDF4<=1.6.0 ?

**Comment by pp-mo** [comment]
Ok for your latest !
I'm totally :+1: on the change to a context manager`dask.config.set` -- I probably should have spotted that + raised it myself.
Still looking forward to some notes about when they fail...

**Comment by trexfeathers** [comment]
# Test results

- Tests: `pytest lib/iris/tests/integration/netcdf/test_thread_safety.py`
- Code base: either [`trexfeathers/netcdf_segs`](https://github.com/trexfeathers/iris/tree/netcdf_segs) or [`v3.4.0`](https://github.com/scitools/iris/tree/v3.4.0)
- Environment specification: either [`trexfeathers/netcdf_segs`](https://github.com/trexfeathers/iris/tree/netcdf_segs) or [`v3.4.0`](https://github.com/scitools/iris/tree/v3.4.0)
- NetCDF4 versions:
  - `netcdf_segs`: `1.6.2`
  - `v3.4.0`: `1.6.0`

(☟ code)  (environment ☞) | `netcdf_segs` | `v3.4.0`
-|-|-
`netcdf_segs` | ✔️ x5 | ✔️ x5
`v3.4.0` | ❌ `test_realise_data` - seg-fault | ✔️ x5

@pp-mo if you'd like us to try harder to trigger thread safety problems I'm open to it. 1 test of 5 isn't particularly 'successful'.

### Notes

- Needed to use `pytest -k` to identify the specific seg-faulting test.
- `test_thread_safety.py` obviously doesn't exist in `v3.4.0` so I had to bring that in manually when running with the `v3.4.0` code base.

File: lib/iris/tests/integration/netcdf/test_thread_safety.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+run.
+
+Token assertions are included after the line that is expected to reveal
+a thread safety problem, as this seems to be good testing practice.
+
+"""
+
+# Import iris.tests first so that some things can be initialised before
+# importing anything else.
+import iris.tests as tests  # isort:skip
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/integration/netcdf/test_thread_safety.py]: I think that's also necessary for setting things like the Matplotlib backend, but we don't need it here.

f6eb131eddbf5ee4080cb3d5c398dea80aca00b5

File: lib/iris/tests/integration/netcdf/test_thread_safety.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+@pytest.fixture
+def get_cubes_from_netcdf():
+    load_dir_path = Path(tests.get_data_path(["NetCDF", "global", "xyt"]))
+    loaded = iris.load(load_dir_path.glob("*"))
+    smaller = CubeList([c[0, 0] for c in loaded])
+    yield smaller
+
+
+def test_load(tiny_chunks, get_cubes_from_netcdf):
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/integration/netcdf/test_thread_safety.py]: 2df627874d6c620559061791589a05e182caf5f8

File: lib/iris/tests/integration/netcdf/test_thread_safety.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        assert save_path.exists()
+
+    yield _func
+
+
+@pytest.fixture
+def get_cubes_from_netcdf():
+    load_dir_path = Path(tests.get_data_path(["NetCDF", "global", "xyt"]))
+    loaded = iris.load(load_dir_path.glob("*"))
+    smaller = CubeList([c[0, 0] for c in loaded])
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/integration/netcdf/test_thread_safety.py]: Thanks, I had missed that extra `Cube`, and I was clearly overzealous trying to get faster running tests - I hadn't bothered to compute the eventual size.

20956582f85d9ab9dd4928549d7e81b98ed9dcda

**Comment by trexfeathers** [comment]
# Discoveries from investigating tests

- More, smaller threads increase the chance of thread collision. So tests using `tiny_chunks` fail more reliably than `_multisource` tests.
- Tests involving `da.store()` rather than `da.compute()` - so those for saving and streaming - NEVER fail, even when streaming from multiple sources. This suggests that Dask's [current] strategy is to serialise the WHOLE operation whenever file writing is involved.
  Good to have those tests in there should the situation ever change in future.

**Comment by trexfeathers** [comment]
Closes #5016

**Review by valeriupredoi — APPROVED** [review]
very very many thanks for tackling this @trexfeathers and everyone else who reviewed and tested - I am approving this from the side of us doing testing with ESMValCore, tests that show this fixes the issue completely; of course, you guys decide when to merge, but it'd be very cool to have this in a new iris release soon :beers: :partying_face:

File: lib/iris/tests/integration/netcdf/test_thread_safety.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+def test_realise_data_multisource(get_cubes_from_netcdf):
+    """Load from multiple sources to force Dask to use multiple threads."""
+    cubes = get_cubes_from_netcdf
+    final_cube = sum(cubes)
+    _ = final_cube.data  # Any problems are expected here.
+    assert not final_cube.has_lazy_data()
+
+
+def test_save(tiny_chunks, save_common):
+    cube = Cube(da.ones(1000))
```
Comment (pp-mo) [inline lib/iris/tests/integration/netcdf/test_thread_safety.py:83]: Should be using a cube large enough to generate >1 chunks.
And we can test that by passing it to the `tiny_chunks` checking method.

**Review by pp-mo — APPROVED** [review]

File: lib/iris/tests/integration/netcdf/test_thread_safety.py (hunk 14 above)
Comment (trexfeathers) [reply lib/iris/tests/integration/netcdf/test_thread_safety.py]: a6b893c2324e070c2a0dab20995c6ae72ba4409b

Comment (pp-mo) [reply lib/iris/tests/integration/netcdf/test_thread_safety.py]: Ok! https://github.com/SciTools/iris/pull/5095/commits/a6b893c2324e070c2a0dab20995c6ae72ba4409b

**Comment by trexfeathers** [comment]
Do you need anything else @pp-mo?

**Comment by pp-mo** [comment]
> Do you need anything else @pp-mo?

No, all good now the tests have passed.
Huge applause ! 👏 👏 🥇

**Pull request merged**